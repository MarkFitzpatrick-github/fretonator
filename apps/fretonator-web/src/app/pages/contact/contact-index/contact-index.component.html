<div class="form__container" #scrollTarget>
  <h1 class="form__title">Share the love <span role="img" aria-label="red heart emoji">❤️</span></h1>
  <p class="form__copy">My goal with
    the Fretonator is to crowdsource feedback and ideas to create the ultimate guitar learning hub that's simple,
    easy and enjoyable to use.</p>
  <p class="form__copy">Noticed a bug? Got an idea for a feature? Want to collaborate? Or just want to say hello? I'd
    love to
    hear from you. Send
    me a message!</p>
  <p class="form__copy form__copy--bold">Let's build the Fretonator together.</p>
  <form [name]="formName"
        data-netlify="true"
        method="POST"
        action="/contact/success"
        netlify-honeypot="bot-field"
        class="form"
        [formGroup]="form"
        (ngSubmit)="onSubmit()">
    <label for="name" class="form__label">
      Your name *
    </label>
    <div class="form__inputGroup">
      <input id="name"
             name="Name"
             type="text"
             formControlName="name"
             class="form__input">
      <span class="form__validIcon" *ngIf="name.valid">
        <app-tick-svg></app-tick-svg>
      </span>
    </div>
    <ng-container *ngIf="name.touched">
      <p class="form__error" *ngIf="name.errors?.required">Enter your name, yo! A fake one will do.</p>
      <p class="form__error" *ngIf="name.errors?.pattern">Please enter a valid name that won't break the form!</p>
      <p class="form__error" *ngIf="name.errors?.maxlength">Come on, your name isn't
        actually {{name.errors?.maxlength.actualLength}}
        characters long!</p>
    </ng-container>
    <label for="email" class="form__label">
      Your contact email
    </label>
    <div class="form__inputGroup">
      <input id="email"
             name="Email"
             type="email"
             formControlName="email"
             class="form__input">
      <span class="form__validIcon" *ngIf="email.value.length > 0 && email.valid">
        <app-tick-svg></app-tick-svg>
      </span>
    </div>
    <ng-container *ngIf="email.touched">
      <p class="form__error" *ngIf="email.errors?.email">An email address isn't required, but if you'd like me to reply
        to your message, please enter a valid one!</p>
    </ng-container>

    <label for="message" class="form__label">
      Your message *
    </label>
    <div class="form__inputGroup">
    <textarea id="message"
              name="Message"
              formControlName="message"
              class="form__input form__input--textArea"></textarea>
      <span class="form__validIcon form__validIcon--textArea" *ngIf="message.valid">
        <app-tick-svg></app-tick-svg>
      </span>
    </div>
    <ng-container *ngIf="message.touched">
      <p class="form__error" *ngIf="message.errors?.required">Go on, give me some feedback!</p>
      <p class="form__error" *ngIf="message.errors?.minlength">Come on, you can write more than that!</p>
    </ng-container>

    <div class="form__honeypot">
      <label for="bot-field">Don't fill this out if you're human!</label>
      <input type="text" id="bot-field" name="bot-field" formControlName="honeypot">
    </div>

    <input type="submit" class="form__submit" formControlName="submit" aria-label="Submit Form"/>
    <p class="form__error form__error--global" *ngIf="formSubmitError">Eeesh! Something went wrong when we tried to
      submit the form. Please
      try again.</p>

  </form>
</div>
