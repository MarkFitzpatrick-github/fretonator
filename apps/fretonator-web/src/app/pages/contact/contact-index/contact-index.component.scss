@import '../../../../styles/mixins';
@import '../../../../styles/typography';

.form__container {
  max-width: $screen-med;
  margin-left: auto;
  margin-right: auto;
  background-color: var(--form-container-bg-color);
  padding: pxToRem($grid-unit * 4);

  @media screen and (min-width: $screen-med) {
    padding: pxToRem($grid-unit * 6) pxToRem($grid-unit * 16);
  }
}

.form__title {
  @include font_large();
  color: var(--form-container-color);
  text-align: left;
  margin-top: pxToRem($grid-unit * 3);
  margin-bottom: pxToRem($grid-unit * 3);
}

.form__copy {
  @include font_bodyCopy();
  color: var(--form-container-color);
  text-align: left;
  margin-bottom: pxToRem($grid-unit * 2);
}

.form__copy--bold {
  font-weight: var(--font-weight-bold);
}

.form {
  margin-top: pxToRem($grid-unit * 2);
  margin-bottom: pxToRem($grid-unit * 2);
}

.form__inputGroup {
  position: relative;
  display: flex;
}

.form__input {
  @include font_bodyCopy();
  border-radius: 0;
  padding: pxToRem($grid-unit) pxToRem($grid-unit * 7)  pxToRem($grid-unit) pxToRem($grid-unit);
  width: 100%;
  display: block;
  border: pxToRem(2) solid var(--form-input-border-color);
  background-color: var(--form-input-bg-color);
  color: var(--form-input-color);
}

.form__input--textArea {
  height: pxToRem($grid-unit * 18);
}

.form__validIcon {
  color: var(--success-bg);
  height: pxToRem($grid-unit * 3);
  position: absolute;
  transform: translateY(50%);
  left: 92%;
}

.form__validIcon--textArea {
  bottom: pxToRem($grid-unit * 3);
}

.form__label {
  @include font_medium();
  font-size: pxToRem(14);
  display: block;
  margin-bottom: pxToRem($grid-unit);
  margin-top: pxToRem($grid-unit * 4);
  color: var(--form-label-color);
}

.form__honeypot {
  display: none;
}

.form__error {
  @include font_bodyCopy();
  background-color: var(--error-bg);
  color: var(--error-fg);
  padding: pxToRem($grid-unit);
  font-weight: var(--font-weight-bold);
  border: pxToRem(2) solid var(--form-input-border-color);
  border-top-width: 0;
}

.form__error--global {
  margin-top: pxToRem($grid-unit * 3);
  margin-bottom: pxToRem($grid-unit * 2);
  border-top-width: pxToRem(2);
}

.form__submit {
  @include hard_button_base();
  margin-top: pxToRem($grid-unit * 6);
  margin-bottom: pxToRem($grid-unit * 3);

  &:disabled {
    opacity: 0.5;
  }
}
