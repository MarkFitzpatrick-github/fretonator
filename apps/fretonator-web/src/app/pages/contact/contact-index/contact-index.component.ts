import { AfterViewInit, Component, ElementRef, ViewChild } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { FormService } from '../form.service';
import { tap } from 'rxjs/operators';
import { Router } from '@angular/router';
import { GlobalService } from '../../../global.service';
import { FormErrorMessages } from '../../../util/constants';

@Component({
  selector: 'app-contact-index',
  templateUrl: './contact-index.component.html',
  styleUrls: ['./contact-index.component.scss']
})
export class ContactIndexComponent implements AfterViewInit {
  @ViewChild('scrollTarget') scrollTarget: ElementRef<HTMLElement>;
  formName = 'Contact';
  formSubmitError = false;
  globalErrors = {
    name: false,
    email: false,
    message: false,
    any: false
  };
  form = new FormGroup({
    honeypot: new FormControl('', [
      Validators.maxLength(0)
    ]),
    name: new FormControl('', [
      Validators.required,
      Validators.maxLength(81),
      Validators.pattern(/[a-zA-Z]+(([',. -][a-zA-Z ])?[a-zA-Z]*)/)
    ]),
    email: new FormControl('', [
      Validators.email
    ]),
    message: new FormControl('', [
      Validators.required,
      Validators.minLength(30)
    ]),
    submit: new FormControl('Submit', [])
  }, {
    updateOn: 'change'
  });

  constructor(private formService: FormService,
              private router: Router,
              private globalService: GlobalService) {
  }

  ngAfterViewInit(): void {
    this.globalService.setScrollTarget(this.scrollTarget.nativeElement);
  }

  onSubmit() {
    this.resetGlobalErrors();

    if (this.form.invalid) {
      this.form.markAllAsTouched();
      this.calculateGlobalErrors();
      return;
    }

    this.formSubmitError = false;
    this.submit.disable();

    const form = {
      Name: this.name.value,
      Email: this.email.value,
      Message: this.message.value,
      'form-name': this.formName,
      'bot-field': this.honeypot.value
    };

    this.formService.submit('/contact/success', form)
      .pipe(tap(() => this.submit.enable()))
      .subscribe(
        () => this.onSuccess(),
        (err) => this.onFail());
  }

  async onSuccess() {
    await (this.router.navigate(['/', 'contact', 'success']));
  }

  onFail() {
    this.formSubmitError = true;
  }

  get name() {
    return this.form.get('name');
  }

  get email() {
    return this.form.get('email');
  }

  get message() {
    return this.form.get('message');
  }

  get submit() {
    return this.form.get('submit');
  }

  get honeypot() {
    return this.form.get('honeypot');
  }

  resetGlobalErrors() {
    this.globalErrors = {
      name: false,
      email: false,
      message: false,
      any: false
    };
  }

  getNameError() {
    if (!this.name.errors) {
      return;
    }
    return true;
  }

  getEmailError() {
    if (!this.email.errors) {
      return;
    }
    return true;
  }

  getMessageError() {
    if (!this.message.errors) {
      return;
    }
    return true;
  }

  calculateGlobalErrors() {
    this.globalErrors = {
      name: this.getNameError(),
      email: this.getEmailError(),
      message: this.getMessageError(),
      any: this.getNameError() || this.getEmailError() || this.getMessageError()
    };
  }

  getErrorMessage(field: string, type: string) {
    return FormErrorMessages[field][type];
  }

}
