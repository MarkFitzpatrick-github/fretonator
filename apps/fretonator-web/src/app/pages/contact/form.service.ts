import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { retry } from 'rxjs/operators';

@Injectable(
)
export class FormService {

  constructor(private httpClient: HttpClient) {
  }

  submit(location, form) {

    const headers = new HttpHeaders()
      .set('Content-Type', 'application/x-www-form-urlencoded');

    const options = {
      headers,
      responseType: 'text' as 'text',
    };

    const formBody = new URLSearchParams();

    Object.entries(form)
      .forEach(([key, value]) =>
        formBody.append(key, value as string));

    return this.httpClient.post(location, formBody.toString(), options)
      .pipe(retry(2));
  }
}
