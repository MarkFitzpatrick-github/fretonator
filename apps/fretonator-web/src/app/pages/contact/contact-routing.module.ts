import { NgModule } from '@angular/core';
import { Route, RouterModule, Routes } from '@angular/router';
import { ContactIndexComponent } from './contact-index/contact-index.component';
import { ContactSuccessComponent } from './contact-success/contact-success.component';
import { ScrollToGuard } from '../../common/scroll-to/scroll-to.guard';

const contact: Route = {
  path: '',
  pathMatch: 'full',
  canActivate: [
    ScrollToGuard
  ],
  component: ContactIndexComponent
};

const success: Route = {
  path: 'success',
  component: ContactSuccessComponent,
  canActivate: [
    ScrollToGuard
  ]
};

const routes: Routes = [contact, success];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class ContactRoutingModule {
}
