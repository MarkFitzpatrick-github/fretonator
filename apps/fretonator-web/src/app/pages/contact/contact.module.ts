import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

import { ContactRoutingModule } from './contact-routing.module';
import { ContactIndexComponent } from './contact-index/contact-index.component';
import { ContactSuccessComponent } from './contact-success/contact-success.component';
import { HttpClientModule } from '@angular/common/http';
import { FormService } from './form.service';


@NgModule({
  declarations: [ContactIndexComponent, ContactSuccessComponent],
  imports: [
    CommonModule,
    ContactRoutingModule,
    ReactiveFormsModule,
    FormsModule,
    HttpClientModule
  ],
  providers: [
    FormService
  ]
})
export class ContactModule { }
