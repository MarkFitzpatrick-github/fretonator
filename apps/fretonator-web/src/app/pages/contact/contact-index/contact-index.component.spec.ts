import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { ContactIndexComponent } from './contact-index.component';
import { Component } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { ContactModule } from '../contact.module';
import { BrowserTestingModule } from '@angular/platform-browser/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

describe('ContactIndexComponent', () => {
  @Component({
    selector: 'app-contact-index-spec',
    template: `
      <app-contact-index></app-contact-index>
    `
  })
  class ContactIndexComponentSpec {
  }

  let component: ContactIndexComponentSpec;
  let fixture: ComponentFixture<ContactIndexComponentSpec>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ContactIndexComponentSpec],
      imports: [
        ContactModule,
        RouterTestingModule
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ContactIndexComponentSpec);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
