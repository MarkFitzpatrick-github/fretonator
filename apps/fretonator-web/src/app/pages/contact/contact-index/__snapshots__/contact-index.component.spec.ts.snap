// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactIndexComponent should create 1`] = `ContactIndexComponentSpec {}`;
